### PR TITLE
Implement context menu entry "Leave this share"

### LINKED ISSUE #5033: Unshare in context menu
- Have an incoming share 
- see that you do not need it 
- how to unshare (on web it is called "leave this share")
--> This should be an option in context menu

### PR DESCRIPTION
**Pull request by allexzander**
This will fix #5033

### CODE REVIEW

File: src/common/syncjournaldb.cpp [hunk 1]
```diff
@@ -49,7 +49,7 @@ Q_LOGGING_CATEGORY(lcDb, "nextcloud.sync.database", QtInfoMsg)
 #define GET_FILE_RECORD_QUERY \
         "SELECT path, inode, modtime, type, md5, fileid, remotePerm, filesize," \
         "  ignoredChildrenRemote, contentchecksumtype.name || ':' || contentChecksum, e2eMangledName, isE2eEncrypted, " \
-        "  lock, lockOwnerDisplayName, lockOwnerId, lockType, lockOwnerEditor, lockTime, lockTimeout, isShared, lastShareStateFetchedTimestmap " \
+        "  lock, lockOwnerDisplayName, lockOwnerId, lockType, lockOwnerEditor, lockTime, lockTimeout, isShared, lastShareStateFetchedTimestmap, isMyShare" \
```
Comment (mgallien) [inline src/common/syncjournaldb.cpp:52]: I have been wondering what `isMyShare` is until I found out that this is to get the information that the file is shared by us or is an outgoing share rather than an incoming share
may I suggest a few other names: `outgoingShare`, `sharedByMe`

File: test/testfolderman.cpp [hunk 2]
```diff
@@ -24,7 +26,108 @@ class TestFolderMan: public QObject
...
+           emit incomingShareDeleted();
+           
+           return reply;
+        });
+
+        QSignalSpy incomingShareDeletedSignal(this, &TestFolderMan::incomingShareDeleted);
+
+        // verify first share gets deleted
+        folderman->leaveShare(fakeFolder.localPath() + firstSharePath);
+        QTest::qWait(100);
```
Comment (mgallien) [inline test/testfolderman.cpp:106]: better do
`QVERIFY(incomingShareDeletedSignal.wait(100));`
that will validate that the signal has been received at least once and will return faster than `100` milliseconds if the signal is emitted before the timeout

File: test/testfolderman.cpp [hunk 3]
```diff
@@ -24,7 +26,108 @@ class TestFolderMan: public QObject
...
+        folderman->leaveShare(fakeFolder.localPath() + "A/B/notsharedwithme_B.data");
+        QTest::qWait(100);
+        QVERIFY(incomingShareDeletedSignal.count() == 1);
+        QVERIFY(fakeFolder.remoteModifier().find("A/B/sharedwithme_B.data"));
+        QVERIFY(fakeFolder.syncOnce());
+        QCOMPARE(fakeFolder.currentLocalState(), fakeFolder.currentRemoteState());
+
+        // verify second share gets deleted
+        folderman->leaveShare(fakeFolder.localPath() + secondSharePath);
+        QTest::qWait(100);
```
Comment (mgallien) [inline test/testfolderman.cpp:122]: same comment

File: src/gui/socketapi/socketapi.h [hunk 4]
```diff
@@ -149,15 +150,13 @@ private slots:
...
+    void
+    sendLockFileCommandMenuEntries(const QFileInfo &fileInfo, Folder *const syncFolder, const FileData &fileData, const SocketListener *const listener) const;
 
     void sendLockFileInfoMenuEntries(const QFileInfo &fileInfo,
-                                     Folder * const syncFolder,
+                                     Folder *const syncFolder,
                                      const FileData &fileData,
-                                     const SocketListener * const listener,
+                                     const SocketListener *const listener,
                                      const SyncJournalFileRecord &record) const;
```
Comment (mgallien) [inline src/gui/socketapi/socketapi.h:160]: looks like unrelated changes

File: src/gui/socketapi/socketapi.h [hunk 5]
```diff
@@ -79,8 +79,7 @@ private slots:
     // Helper structure for getting information on a file
     // based on its local path - used for nearly all remote
     // actions.
-    struct FileData
-    {
+    struct FileData {
```
Comment (mgallien) [inline src/gui/socketapi/socketapi.h:82]: class and struct should have the starting accolade into next line

File: src/gui/socketapi/socketapi.h [hunk 6]
```diff
@@ -171,7 +170,7 @@ private slots:
 
     /// Direct Editing
     Q_INVOKABLE void command_EDIT(const QString &localFile, OCC::SocketListener *listener);
-    DirectEditor* getDirectEditorForLocalFile(const QString &localFile);
+    DirectEditor *getDirectEditorForLocalFile(const QString &localFile);
```
Comment (mgallien) [inline src/gui/socketapi/socketapi.h:173]: unrelated change

File: test/testfolderman.cpp (hunk 2 above)
Comment (allexzander) [reply test/testfolderman.cpp:106]: @mgallien no, the test then fails

Comment (mgallien) [reply test/testfolderman.cpp:106]: yes because the signal has been emitted before you call `QTest::qWait(100);` so calling `QVERIFY(incomingShareDeletedSignal.wait(100));` would fail
I would just remove the call to `QTest::qWait` because it is not needed

File: test/testfolderman.cpp [hunk 7]
```diff
@@ -24,7 +26,105 @@ class TestFolderMan: public QObject
...
+        // verify no share gets deleted
+        folderman->leaveShare(fakeFolder.localPath() + "A/B/notsharedwithme_B.data");
+        QVERIFY(incomingShareDeletedSignal.count() == 1);
+        QVERIFY(fakeFolder.remoteModifier().find("A/B/sharedwithme_B.data"));
+        QVERIFY(fakeFolder.syncOnce());
+        QCOMPARE(fakeFolder.currentLocalState(), fakeFolder.currentRemoteState());
+
+        // verify second share gets deleted
+        folderman->leaveShare(fakeFolder.localPath() + secondSharePath);
+        QVERIFY(incomingShareDeletedSignal.count() == 2);
```
Comment (mgallien) [inline test/testfolderman.cpp:120]: it would be better to use `QCOMPARE` for that as in case of a failure, you clearly see the expected value and the real value
see other lines similar that use `QVERIFY` to compare real and expected values
